aidial_adapter_bedrock/llm/converse/input.py

```diff
@@ -0,0 +1,293 @@
+import json
+from typing import List, Set, Tuple, assert_never
+
+from aidial_sdk.chat_completion import FunctionCall as DialFunctionCall
+from aidial_sdk.chat_completion import Message as DialMessage
+from aidial_sdk.chat_completion import (
+    MessageContentImagePart,
+    MessageContentTextPart,
+)
+from aidial_sdk.chat_completion import Role as DialRole
+from aidial_sdk.chat_completion import ToolCall as DialToolCall
+from aidial_sdk.exceptions import RuntimeServerError
+
+from aidial_adapter_bedrock.dial_api.request import ToolsConfig
+from aidial_adapter_bedrock.dial_api.resource import (
+    AttachmentResource,
+    URLResource,
+)
+from aidial_adapter_bedrock.dial_api.storage import FileStorage
+from aidial_adapter_bedrock.llm.converse.types import (
+    ConverseContentPart,
+    ConverseMessage,
+    ConverseRole,
+    ConverseTextPart,
+    ConverseToolResultPart,
+    ConverseTools,
+    ConverseToolUsePart,
+)
+from aidial_adapter_bedrock.llm.errors import ValidationError
+from aidial_adapter_bedrock.utils.list import group_by
+from aidial_adapter_bedrock.utils.list_projection import ListProjection
+
+
+def to_converse_role(role: DialRole) -> ConverseRole:
+    """
+    Converse API accepts only 'user' and 'assistant' roles
+    """
+    match role:
+        case (
+            DialRole.USER | DialRole.TOOL | DialRole.FUNCTION | DialRole.SYSTEM
+        ):
+            return ConverseRole.USER
+        case DialRole.ASSISTANT:
+            return ConverseRole.ASSISTANT
+        case _:
+            assert_never(role)
+
+
+def to_converse_tools(tools_config: ToolsConfig) -> ConverseTools:
+    tools = []
```

Comment:
`tools: list[ConverseToolSpec] = []`

otherwise, typechecker doesn't validate that the followup lines actually append `ConverseToolSpec`-compatible dicts to the list.

Reply:
Fixed